src/Driver.cpp

```diff
@@ -154,78 +156,94 @@ int Driver::getFileDescriptor() const
 }
 bool Driver::isValid() const { return m_stream; }
 
-void Driver::openURI(std::string const& uri)
-{
+void Driver::openURI(std::string const& uri) {
     // Modes:
     //   0 for serial
     //   1 for TCP
     //   2 for UDP
     //   3 for UDP server
     //   4 for file (either Unix sockets or named FIFOs)
     int mode_idx = -1;
-    char const* modes[6] = { "serial://", "tcp://", "udp://", "udpserver://", "file://", "test://" };
-    for (int i = 0; i < 6; ++i)
-    {
-        if (uri.compare(0, strlen(modes[i]), modes[i]) == 0)
-        {
+    char const* modes[6] = { "serial://", "tcp://", "udp://",
+                             "udpserver://", "file://", "test://" };
+    for (int i = 0; i < 6; ++i) {
+        if (uri.compare(0, strlen(modes[i]), modes[i]) == 0) {
             mode_idx = i;
             break;
         }
     }
 
-    if (mode_idx == -1)
+    if (mode_idx == -1) {
         throw std::runtime_error("unknown URI " + uri);
-
+    }
     string device = uri.substr(strlen(modes[mode_idx]));
 
     // Find a :[additional_info] marker
     string::size_type marker = device.find_last_of(":");
+
     int additional_info = 0;
-    if (marker != string::npos)
-    {
-        additional_info = boost::lexical_cast<int>(device.substr(marker + 1));
+    SerialConfiguration serial_config;
+    bool has_serial_config = false;
+
+    if (marker != string::npos) {
```

Comment:
The original "only everything after the last : and it has to be a number" doesn't hold up anymore really. I think it would be best to make URI parsing per-mode, just creating helpers for the common cases.

This would also help testing, as we can add unit tests for the parsing methods.

Reply:
I agree :+1: